tt-rss service: init at 16.3

zohl opened the pull request:
This patch adds Tiny Tiny RSS service.
I tested only basic functionality (i.e. without Sphinx, email notifications etc.), but added the corresponding options.

The service does only half a job (database initialization and feed updates). In order to access the interface there should be a web-server with php application server and a database with already created user.
Here is example how to do it with `nginx` and `php-fpm`:
https://gist.github.com/zohl/05c1e5830dc0c898a8808204fedbf76b

Since db initialization scripts are quite aggressive (they purge all the data and create the schema from scratch), they will be called only if the database contains no data. There is a little chance that the output format of `psql` or `mysql` will be different and the things will go wrong. But even so, the only non-desirable effect would be missed schema initialization.
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [X] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

zohl committed:
tt-rss service: init at 16.3

groxxda commented on nixos/modules/misc/ids.nix:
```diff
@@ -270,6 +270,8 @@
       toxvpn = 247;
 
+      tt-rss = 249;
+
```
Please also add a comment below to reserve the corresponding gid

groxxda commented on nixos/modules/services/web-apps/tt-rss.nix:
```diff
@@ -0,0 +1,521 @@
+
+    services.tt-rss = {
+
+      enable = mkOption {
```
Maybe use mkEnableOption

groxxda commented on nixos/modules/services/web-apps/tt-rss.nix:
```diff
@@ -0,0 +1,521 @@
+        '';
+      };
+
+      user = mkOption {
```
Is there a good reason to make this configurable? Otherwise please drop the option

groxxda commented on nixos/modules/services/web-apps/tt-rss.nix:
```diff
@@ -0,0 +1,521 @@
+          default = ["ttrss" "delta"];
+          description = ''
+            Index name in Sphinx configuration. You can specify multiple indexes
+            as a comma-separated string.
```
Could use rewording because this is a list and not a string of csv

groxxda commented on nixos/modules/services/web-apps/tt-rss.nix:
From the description this should be an int to allow setting arbitrary lifetimes other than 1 day

groxxda commented on nixos/modules/services/web-apps/tt-rss.nix:
Wording regarding list vs csv like above

groxxda commented on nixos/modules/services/web-apps/tt-rss.nix:
Is this setting useful? I think this is determined by the output of this module and thus overriding it makes little sense

groxxda commented on nixos/modules/services/web-apps/tt-rss.nix:
Is 777 necessary? :disappointed:

groxxda commented on nixos/modules/services/web-apps/tt-rss.nix:
This probably really requires MySQL or pgsql not only 'after'. Also you could skip one dependency depending on cfg.database.type

groxxda commented:
It would be awesome if you find the time to add a documentation chapter with the sample configuration from your gist :+1: 
Also it may be better to check in preStart if the directory exists and if it does skip parts of the  initialization.
And maybe symlink the config instead of copying - that way it is read only in the nixstore and nobody is tempted to edit it an gets frustrated when it's overwritten on the next nixos rebuild

zohl commented:
@groxxda, thank you very much for your review! Looks like I shall pay more attention to the details before commit :)
I updated the PR according to your recomendations, but the only thing I would disagree with is the checking mechanism in `preStart` script.

The difference between the directory containing tt-rss stuff and the database is that the former can be easily restored while the latter don't. The one can naturally assume that deleting the directory won't affect the feeds (as they stored physically in different location), but in reality the next `nixos-rebuild` will wipe this data. And this would be very frustrating. I might be a paranoid a little bit (or just too lazy to make a backup before system update), but I prefer to keep more complex check in order to prevent such a scenario.

Also there is a change in `user` parameter. We can avoid granting `777` to files, but both the service and the web-application should have write access to the directory. I think the simplest way is to run the service and the application under the same account (like `nginx`or `http`). No more extra users and `777`s :)

zohl commented:
And yes, I have got some time to contribute to the documentation. AFAIK, the wiki is read only and the appropriate place should be `nixpkgs/nixos/doc/manual/configuration/`, but there are only articles that deal with "serious" topics like networking and kernels. Is it really OK to push an article on a small service there or is there more appropriate place?

zohl commented:
>I think the manual would be the right place, but I'm also unsure whether we want an article for every small service there.

I guess it would be reasonable to create a separate section for such small services. Moreover the other contributors can also benefit from it.


>Of course it would be even better if the module could be made as self-contained as possible:

Sure thing! I didn't think about it, but since there is a way for nginx and phpfpm (not sure for the databases though) it should be definitely employed. Thanks for the idea!


>One more thing: Why did you write a derivation for the config file? 

I do not know any other way to put a file in the nix store :)
Is there a better way?
 

PS. Sorry for being absent, I'll fix it on the upcoming weekends.

gebner commented:
I'm running this personally now, and it seems to work fine so far.  I had to fix one import issue though:
```php
define('SINGLE_USER_MODE', '${boolToString cfg.singleUserMode}');
```
Do not quote booleans in PHP.  This unintentionally disables authentication because PHP treats `'false'` as true...
```
$ echo '<?php if ("false") { echo("then branch\n"); } else { echo("else branch\n"); }' | nix-shell -p php --run php
then branch
```

zohl committed:
phpfpm service: restructured pool configuration

zohl commented:
@gebner, thanks for the point! I've fixed all boolean options.

I also added options for phpfpm pool and nginx virtual host with default configurations, so tt-rss can be started by the following (pseudo)code:
```
services.nginx = {
  enable = true;
};

services.tt-rss = {
  enable = true;
  selfUrlPath = "http://localhost";
  database = { ... };
};

services.postgresql = { ... };
```
Since there is not so much non-trivial stuff, I guess there is no need to add notes into the manual.
And as the virtual hosts were employed here, this PR is requires #15862.

garbas commented on pkgs/servers/tt-rss/default.nix:
```diff
@@ -0,0 +1,27 @@
+    description = "Web-based news feed (RSS/Atom) aggregator";
+    license = licenses.gpl2Plus;
+    homepage = http://tt-rss.org;
+    platforms = platforms.all;
```
can you add yourself as a maintainer?

zohl replied:
Sure, I've updated the PR.

the-kenny commented:
Please note that I disabled the `nginx.virtualHost`-feature in f8ea8c7197943b1cc637bd8a796398111185bd1b. 

We can re-enable it when https://github.com/NixOS/nixpkgs/pull/15862 is merged.

zohl commented:
@the-kenny, got it, thanks!

bjornfor commented:
I dislike that this removes ```services.phpfpm.poolConfigs``` without an explanation why (in the commit message). Nor does it provide a way to guide users towards the new interface:

```
$ nixos-rebuild ...
building Nix...
building the system configuration...
error: The option `services.phpfpm.poolConfigs' defined in `/etc/nixos/modules/nextcloud.nix' does not exist.
(use ‘--show-trace’ to show detailed location information)
```

IMHO, configuration.nix files should *work* in both stable and unstable. Deprecate an interface in release N+1, remove the interface in release N+2.

bjornfor commented:
IMHO, we should either revert the phpfpm interface change or bring back the poolConfigs interface (and mark it as deprecated). Opinions?

bjornfor commented:
CC @rickynils (who added phpfpm module to NixOS).
